src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs

```diff
@@ -149,117 +170,156 @@ public override void Dispose()
-                    packet = new SNIPacket(headerSize: 0, dataSize: _bufferSize);
-                    packet.ReadFromStream(_stream);
-
-                    if (packet.Length == 0)
+                    packet = null;
+                    try
+                    {
+                        packet = new SNIPacket(headerSize: 0, dataSize: _bufferSize);
+                        packet.ReadFromStream(_stream);
+
+                        if (packet.Length == 0)
+                        {
+                            errorPacket = packet;
+                            packet = null;
+                            var e = new Win32Exception();
+                            SqlClientEventSource.Log.SNITrace("<sc.SNI.SNINpHandle.Receive |SNI|ERR> packet length is 0.");
+                            return ReportErrorAndReleasePacket(errorPacket, (uint)e.NativeErrorCode, 0, e.Message);
+                        }
+                    }
+                    catch (ObjectDisposedException ode)
+                    {
+                        errorPacket = packet;
+                        packet = null;
+                        SqlClientEventSource.Log.SNITrace("<sc.SNI.SNINpHandle.Receive |SNI|ERR> ObjectDisposedException message = {0}.", ode.Message);
+                        return ReportErrorAndReleasePacket(errorPacket, ode);
+                    }
+                    catch (IOException ioe)
                     {
                         errorPacket = packet;
                         packet = null;
-                        var e = new Win32Exception();
-                        return ReportErrorAndReleasePacket(errorPacket, (uint)e.NativeErrorCode, 0, e.Message);
+                        SqlClientEventSource.Log.SNITrace("<sc.SNI.SNINpHandle.Receive |SNI|ERR> IOException message = {0}.", ioe.Message);
+                        return ReportErrorAndReleasePacket(errorPacket, ioe);
                     }
+                    return TdsEnums.SNI_SUCCESS;
+                }
+            }
+            finally
+            {
+                SqlClientEventSource.Log.ScopeLeaveEvent(scopeID);
+            }
+        }
+
+        public override uint ReceiveAsync(ref SNIPacket packet)
+        {
+            long scopeID = SqlClientEventSource.Log.SNIScopeEnter("<sc.SNI.SNINpHandle.ReceiveAsync |SNI|SCOPE>");
+            try
+            {
+                SNIPacket errorPacket;
+                packet = new SNIPacket(headerSize: 0, dataSize: _bufferSize);
+
+                try
+                {
+                    packet.ReadFromStreamAsync(_stream, _receiveCallback);
+                    return TdsEnums.SNI_SUCCESS_IO_PENDING;
                 }
                 catch (ObjectDisposedException ode)
                 {
                     errorPacket = packet;
                     packet = null;
+                    SqlClientEventSource.Log.SNITrace("<sc.SNI.SNINpHandle.ReceiveAsync |SNI|ERR> ObjectDisposedException message = {0}.", ode.Message);
```

Comment:
this was only to capture and clarify where the exception has happened. it could be only mentioned the name of the class, but I followed the general pattern to include the exception message.